agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs

```diff
@@ -1 +1,347 @@
+                        "b".to_string() => Schema::Document(Document { keys:
+                            map! {
+                                "x".to_string() => Schema::Atomic(Atomic::Boolean),
+                                "y".to_string() => Schema::Atomic(Atomic::Integer)
+                            },
+                            required: set!("x".to_string(), "y".to_string()),
+                            ..Default::default()
+                        }),
+                        "partition_one".to_string() => Schema::Atomic(Atomic::Double),
+                    },
+                    required: set!("a".to_string(), "b".to_string(), "partition_one".to_string()),
+                    ..Default::default()
+                }),
+                "partition_two".to_string() => Schema::Atomic(Atomic::Double)
+            },
+            required: set!("foo".to_string(), "bar".to_string()),
+            ..Default::default()
+        })
+    );
+}
+
+mod documents {
+    use super::*;
+
+    test_derive_stage_schema!(
+        empty,
+        expected = Ok(Schema::Document(Document::default())),
+        input = r#"{"$documents": []}"#
+    );
+
+    test_derive_stage_schema!(
+        singleton,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "a".to_string() => Schema::Atomic(Atomic::Integer)
+            },
+            required: set!("a".to_string()),
+            ..Default::default()
+        })),
+        input = r#" {"$documents": [{"a": 1}]}"#
+    );
+
+    test_derive_stage_schema!(
+        multiple_documents,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "a".to_string() => Schema::AnyOf(set!(
+                    Schema::Atomic(Atomic::Integer),
+                    Schema::Atomic(Atomic::String),
+                    Schema::Document(Document {
+                        keys: map! {
+                            "b".to_string() => Schema::Document(Document {
+                                keys: map!{
+                                    "c".to_string() => Schema::Atomic(Atomic::Boolean)
+                                },
+                                required: set!("c".to_string()),
+                                ..Default::default()
+                            })
+                        },
+                        required: set!("b".to_string()),
+                        ..Default::default()
+                    })
+                )),
+                "b".to_string() => Schema::AnyOf(set!(
+                    Schema::Atomic(Atomic::Null),
+                    Schema::Atomic(Atomic::Integer)
+                ))
+            },
+            required: set!("a".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$documents": [
+                     {"a": 1, "b": 2},
+                     {"a": "yes", "b": null},
+                     {"a": {"b": {"c": true}}}
+        ]}"#
+    );
+}
+
+mod facet {
```

Comment:
To make these tests sensible / follow existing examples, I thought it might be best to just wait for more stages to be implemented. I left them here though so that the future implementer doesn't need to copy them in.

Reply:
I'm a fan of this